Remove deprecated Vmdb::Config class

tadeboro opened the pull request:
This pull request removes deprecated `Vmdb::Config` class and modifies all its consumers.

This is initial attempt at silencing deprecation warnings from #17953 

@miq-bot add_label wip

tadeboro commented:
@miq-bot add_label wip

djberg96 commented:
:+1:

Fryguy commented on spec/lib/vmdb/config/validator_spec.rb:
```diff
@@ -5,11 +5,6 @@
     end
 
-    it "with a VMDB::Config" do
-      validator = described_class.new(VMDB::Config.new("vmdb"))
```
If we are removing this, then we should also remove the support for it: https://github.com/ManageIQ/manageiq/blob/78f10b36dabbf67ed05d6553b9231767e165fa99/lib/vmdb/config/validator.rb#L10-L11

Otherwise, this test should be changed to do the OpenStruct trick like you did in the UI code.  However, I'd prefer just dropping it, and changing the UI to call the validator with the inner config.

tadeboro replied:
I removed support for `VMDB::Config` from both validator and activator.

Fryguy commented on app/models/miq_server/configuration_management.rb:
```diff
@@ -2,12 +2,9 @@ module MiqServer::ConfigurationManagement
-    end
+  def get_config(*_ignored)
+    config = is_local? ? ::Settings.to_hash : Vmdb::Settings.for_resource(self)
+    OpenStruct.new("config" => config)
```
is it possible to kill this OpenStruct.new line?

Fryguy commented on app/models/miq_server/configuration_management.rb:
```diff
@@ -2,16 +2,11 @@ module MiqServer::ConfigurationManagement
-      VMDB::Config.for_resource(type, self)
-    end
+  def config
+    (is_local? ? ::Settings : Vmdb::Settings.for_resource(self)).to_hash
```
Should this be `::Settings.to_hash`?

tadeboro replied:
WE call `to_hash` a few characters to the right.

Fryguy replied:
Ah... I didn't see the outer parens wrapped both.

carbonin commented on app/models/miq_server/configuration_management.rb:
```diff
@@ -2,16 +2,11 @@ module MiqServer::ConfigurationManagement
-    else
-      VMDB::Config.for_resource(type, self)
-    end
+  def config
```
As much as rubocop doesn't like `get_*` methods I think just `config` might be a bit too likely to conflict. Plus it's not very descriptive. Maybe something like `server_config` or `server_settings`?

tadeboro replied:
This way my train of thought exactly when rubocop complained. But after I checked the actual usage, `config` did not seem that bad anymore, since most of the time, this is called as `MiqServer.my_server.config` and the rest of the calls being in the form of `MiqServer.find(...).config`. I think that having `MiqServer.my_server.server_config` or `MiqServer.find(...).server_settings` does not add much clarity (and that duplicated server in the case of `my_server` just looks bad) over bare `config`.

If my argument is not strong enough, I will replace `config`/`set_config` with `server_settings`/`set_server_settings`.

carbonin commented on lib/vmdb/config/activator.rb:
```diff
@@ -1,17 +1,10 @@
 module VMDB
-  class Config
+  module Config
```
Why change this from a class to a module? Couldn't we also remove the `Config` namespace from these, maybe put them under `/lib/vmdb/settings/`?

As it stands you'll need to change https://github.com/ManageIQ/manageiq/blob/master/lib/vmdb/settings.rb#L51

carbonin commented on lib/vmdb/config/validator.rb:
```diff
@@ -1,17 +1,10 @@
 module VMDB
-  class Config
+  module Config
```
Same concern as the `Activator`. We should either keep this a class and change the namespacing, or at least change https://github.com/ManageIQ/manageiq/blob/master/lib/vmdb/settings.rb#L55

carbonin commented on spec/models/miq_schedule_worker/runner_spec.rb:
```diff
@@ -533,13 +530,6 @@
-      stub_server_configuration(
-        :ems_refresh => {
-          :refresh_interval => 24.hours,
-          :scvmm            => {:refresh_interval => 15.minutes}
```
I think this (or something like it) is still needed. The only reason this works still is that these happen to be the values that are set in [settings.yml](https://github.com/ManageIQ/manageiq/blob/master/config/settings.yml#L172-L174). I would prefer to make this test use something other than the defaults, but if the defaults in settings.yml are changed for some reason this test will fail.

tadeboro replied:
I will check the tests that were using this method and find the appropriate replacement.

carbonin replied:
Thanks, it looks like this one at least is a case of a poorly framed test, but given that the PR is green it seems like that might be the case in quite a few places :worried:

tadeboro replied:
My plan was to find proper replacement for stub method and then "fix" the tests by using some intentionally ridiculous values in assertions.

tadeboro replied:
I went through tests again and replaced `stub_server_configuration` calls with `stub_settings` calls that should achieve the same effect. This caused three tests to start failing, two of which could be easily fixed. I marked last remaining failing test with a `skip` call and should be either changed to test the method behavior or the method should be fixed.

carbonin commented on app/models/miq_server/ntp_management.rb:
```diff
@@ -11,7 +11,7 @@ def ntp_reload
     # Bust the settings cache allowing this worker to apply any recent changes made by another (UI) worker
     Vmdb::Settings.reload!
-    ntp_settings = get_config("vmdb").config[:ntp]
+    ntp_settings = config[:ntp]
```
My issue with the `config` method naming really came up because of this.

I found it plausible for a local variable or attribute to be called `config` and it took me a bit to realize you had renamed `get_config`. I'm more concerned with the implicit `self` target than the explicit target case (which, as you said, makes this more clear).

tadeboro replied:
Uhg, jeah. I see your point now. But maybe doing explicit `self.config` would do? Otherwise, I will go and replace it as discussed before.

Fryguy replied:
If we use settings as the method name, then maybe the collision is less likely?  Otherwise, I'm fine with self.settings, though rubocop will likely complain.

carbonin commented on spec/lib/vmdb/util_spec.rb:
```diff
@@ -1,7 +1,8 @@
   context ".http_proxy_uri" do
     it "without config settings" do
-      stub_server_configuration({})
+      skip(".http_proxy_uri does not handle missing settings yet")
```
What was the failure here? I'd rather fix this in this PR than add a `skip`

tadeboro replied:
I was about calling `[]` on `nil`, since `.http_proxy_uri` assumes some fields are always present in settings. I do not remember exactly what failed, but I think things fell apart at https://github.com/ManageIQ/manageiq/blob/master/lib/vmdb/util.rb#L6

tadeboro commented:
@Fryguy @carbonin I went ahead and renamed `get_config`/`set_config` methods to `settings`/`settings=` as @Fryguy suggested. As for the skipped test, I modified the settings stub call in order to make the test pass. I kept modifications down to a minimum in order to keep test simple.

carbonin commented on app/models/miq_server/configuration_management.rb:
```diff
@@ -2,17 +2,12 @@ module MiqServer::ConfigurationManagement
-    else
-      VMDB::Config.for_resource(type, self)
-    end
+  def settings
```
I'm not sure I understand the concern. This is a `settings` method on `MiqServer` only, why would it matter if there is a `Zone#settings`? Or are you just confused because they would point to different things?

Fryguy replied:
yeah they would be different, but both (+MiqRegion) are "things from which you can get Settings"

Fryguy replied:
there is no code conflict...more of a conceptual conflict.  Again, it might be ok.

carbonin commented on app/models/miq_server/configuration_management.rb:
```diff
@@ -2,17 +2,12 @@ module MiqServer::ConfigurationManagement
-      VMDB::Config.for_resource(type, self)
-    end
+  def settings
+    (is_local? ? ::Settings : Vmdb::Settings.for_resource(self)).to_hash
```
Actually, now that @Fryguy mentioned `#settings_for_resource`, we could use that in place of `Vmdb::Settings.for_resource(self)` here.

Fryguy commented on lib/vmdb/settings/validator.rb:
```diff
@@ -104,7 +93,7 @@ def session(data)
           end
 
-          if data.timeout == 0
+          if data.timeout.kind_of?(Integer) && data.timeout.zero?
```
I kind of preferred the original code...more concise and cleaner.  Why did you change?  If because of Rubocop, we can just ignore it.

tadeboro replied:
Yes, this was done to please the Rubocop.

Fryguy commented:
Overall, I am good with this at this point.  I don't think my minor concerns warrant holding it up.  This must be merged with the UI side, so if that one is ready to go, then I approve.

carbonin commented on app/models/miq_server/configuration_management.rb:
```diff
@@ -2,17 +2,12 @@ module MiqServer::ConfigurationManagement
-  def set_config(config)
-    config = config.config if config.respond_to?(:config)
-    add_settings_for_resource(config)
+  def settings=(new_settings)
```
After looking at the UI PR, I'm wondering if we can just remove this method entirely and use `#add_settings_for_resource` as this method doesn't add any additional logic now.

tadeboro commented:
@Fryguy @carbonin I killed the `settings=` method with the last commit. If this is OK, I will go and update the UI repo to get it back in sync with core.

carbonin commented:
Yeah, I think the `MiqServer#settings` is okay. If you just undo those rubocop changes @Fryguy pointed out then this one looks good to go.

tadeboro commented:
I reverted some of the changes that I made in order to please the Rubocop (which means that miq-bot should start complaining really soon;).

tadeboro commented:
@miq-bot add_label wip

tadeboro commented:
@miq-bot remove_label wip

carbonin commented:
@tadeboro can we leave this WIP until https://github.com/ManageIQ/manageiq-ui-classic/pull/4643 is merged?

tadeboro commented:
@carbonin UI PR has been merged. Removing WIP.

@miq-bot remove_label wip
